File: 090_Subsets_II/090_1.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        int size = nums.length;
        if (size == 0) {
            res.add(new ArrayList<>());
            return res;
        }
        Arrays.sort(nums);
        List<Integer> tmp = new ArrayList<>();
        dfs(nums, res, tmp, 0);
        return res;
    }

    private void dfs(int[] nums, List<List<Integer>> res, List<Integer> tmp, int idx) {
        res.add(new ArrayList<>(tmp));
        for (int i = idx; i < nums.length; i++) {
            if (i > idx && nums[i] == nums[i-1]) continue;
            tmp.add(nums[i]);
            dfs(nums, res, tmp, i + 1);
            tmp.remove(tmp.size() - 1);
        }
    }
}

File: 090_Subsets_II/090_1.py
from typing import List

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        tmp = []

        def dfs(tmp, index):
            res.append(tmp)
            for i in range(index, len(nums)):
                if i > index and nums[i-1] == nums[i]:
                    continue
                dfs(tmp + [nums[i]], i + 1)

        dfs(tmp, 0)
        return res